monolith/src/monolith.rs

```diff
@@ -0,0 +1,226 @@
+//! The Monolith-31 permutation.
+//! With significant inspiration from https://extgit.iaik.tugraz.at/krypto/zkfriendlyhashzoo/
+
+extern crate alloc;
+
+use alloc::borrow::ToOwned;
+use alloc::vec;
+use alloc::vec::Vec;
+use core::iter;
+use core::marker::PhantomData;
+
+use p3_field::{AbstractField, PrimeField32, PrimeField64};
+use p3_mds::MdsPermutation;
+use p3_mersenne_31::Mersenne31;
+use sha3::digest::{ExtendableOutput, Update};
+use sha3::{Shake128, Shake128Reader};
+
+use crate::util::get_random_u32;
+
+// The Monolith-31 permutation over Mersenne31.
+pub struct MonolithMersenne31<Mds, const WIDTH: usize, const NUM_ROUNDS: usize>
+where
+    Mds: MdsPermutation<Mersenne31, WIDTH>,
+{
+    // TODO: if possible, replace with [[Mersenne31; WIDTH]; NUM_ROUNDS - 1]]
```

Comment:
Hm maybe change it to a const generic for the number of "complete"/"full" rounds? Otherwise we'd need `generic_const_exprs`, which seems too buggy for now. (A Vec as you have it also seems reasonable though.)

Reply:
Hm good idea; it's a little gross but I prefer the `NUM_FULL_ROUNDS` version